sky/spot/recovery_strategy.py

```diff
@@ -142,39 +142,57 @@ def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
+                    job_lib.JobStatus.INIT,
+                    job_lib.JobStatus.PENDING,
+            ]) and job_checking_retry_cnt < spot_utils.MAX_JOB_CHECKING_RETRY:
+                job_checking_retry_cnt += 1
+                cluster_status, _ = backend_utils.refresh_cluster_status_handle(
+                    self.cluster_name, force_refresh=True)
+                if cluster_status != global_user_state.ClusterStatus.UP:
+                    # The cluster can be preempted before the job is launched.
+                    # In this case, we will terminate the cluster and retry
+                    # the launch.
+                    retry_launch = True
+                    break
                 # Wait the job to be started
                 status = spot_utils.get_job_status(self.backend,
                                                    self.cluster_name)
-                while (status is None or status in [
-                        job_lib.JobStatus.INIT,
-                        job_lib.JobStatus.PENDING,
-                ]):
-                    time.sleep(spot_utils.JOB_STARTED_STATUS_CHECK_GAP_SECONDS)
-                    status = spot_utils.get_job_status(self.backend,
-                                                       self.cluster_name)
-                launch_time = spot_utils.get_job_timestamp(self.backend,
-                                                           self.cluster_name,
-                                                           get_end_time=False)
-                return launch_time
-
-            if max_retry is not None and retry_cnt >= max_retry:
-                # Retry forever if max_retry is None.
-                if raise_on_failure:
-                    with ux_utils.print_exception_no_traceback():
-                        raise exceptions.ResourcesUnavailableError(
-                            'Failed to launch the spot cluster after '
-                            f'{max_retry} retries.')
-                else:
-                    return None
-            gap_seconds = backoff.current_backoff()
-            logger.info(
-                f'Retrying to launch the spot cluster in {gap_seconds:.1f} '
-                'seconds.')
-            time.sleep(gap_seconds)
+                time.sleep(spot_utils.JOB_STARTED_STATUS_CHECK_GAP_SECONDS)
+                status = spot_utils.get_job_status(self.backend,
+                                                   self.cluster_name)
+            if retry_launch:
+                self.terminate_cluster()
+                gap_seconds = backoff.current_backoff()
+                logger.info(
+                    'Failed to check the job status, probably due to the '
+                    'preemption or job submission process failure. Retrying '
+                    f'to launch the cluster in {gap_seconds:.1f} seconds.')
+                time.sleep(gap_seconds)
+                continue
+            launch_time = spot_utils.get_job_timestamp(self.backend,
```

Comment:
Cluster may get preempted right before/during this call? Should we move this to L182 and do it only if `status != None`, and make it protected by the retry logic there? (It may be better to combine these 2 funcs into `get_job_status_and_timestamp(...) -> Optional[Tuple[..., ...]]`, but we can leave that as a TODO.)

Reply:
Good catch! I refactor the retry logic, and now this will imediately follow the job status transition (if the retry happens, the while loop will continue without reaching this line). The preemption can still happen between the two operations but it should be rare (added a TODO).

Reply:
My local branch moved this into a try-catch. But I lost the logs to verify whether the except block would be triggered. But an overnight run doesn't return any FAILED_CONTROLLER anymore. Seems safer to add?

E.g., put this  after L186?  Feel free to push back if you see a downside.

Reply:
Ahh, that is a great idea! After moving the launch time into the job status checking while loop, we can reuse the retry logic for job checking to fix any transient issue that can caused by that command. Just moved the line. Thanks!